camel/agents/chat_agent.py

```diff
@@ -344,33 +333,20 @@ def step(
 
         return ChatAgentResponse(output_messages, self.terminated, info)
 
-    def preprocess_messages(
-            self,
-            messages: List[ChatRecord]) -> Tuple[List[OpenAIMessage], int]:
-        r"""Truncate the list of messages if message window is defined and
-        the current length of message list is beyond the window size. Then
-        convert the list of messages to OpenAI's input format and calculate
-        the number of tokens.
-
-        Args:
-            messages (List[ChatRecord]): The list of structs containing
-                information about previous chat messages.
+    def preprocess_messages(self) -> Tuple[List[OpenAIMessage], int]:
+        r""" Convert the messages in the agent memory to OpenAI's input format
+        and calculate the number of tokens.
 
         Returns:
             tuple: A tuple containing the truncated list of messages in
                 OpenAI's input format and the number of tokens.
         """
-
-        if (self.message_window_size
-                is not None) and (len(messages) > self.message_window_size):
-            messages = [ChatRecord('system', self.system_message)
-                        ] + messages[-self.message_window_size:]
-
-        openai_messages: List[OpenAIMessage]
-        openai_messages = [record.to_openai_message() for record in messages]
+        records = self.memory.retrieve()
```

Comment:
Conceptually `retrieve` must have a `key` parameter. Reading from the head of the memory is called `peek`. Peek is conceptually parameterless. Thus, rename this method to peek.

Reply:
Okay, I've got the idea. I think the method must be called `get_records_for_llm_context`. @lightaime we need a clear name that is self-explanatory. The current naming is too unspecific.

Reply:
> Okay, I've got the idea. I think the method must be called `get_records_for_llm_context`

It's clear but I think including "for_llm_context" is unnecessary because it's the only purpose. However, I want to emphasize this function is just the default way to retrieve memory. How about simple `get_records`? Besides, If we want to get records by a string, how do we call that? `get_records_by_str`?

> we need a clear name that is self-explanatory. The current naming is too unspecific.

Highly agree.

Reply:
For me get_records is unclear. It is a parameterless function. What recored does it return? The last? The last 5? All records? Name questions and no answers. We must emphasize that this function returns messages that are tokenizable into less or equal of the token limit of some specific model. Suggest your name to account for this.

Reply:
What is "getting records by a string"? It is completely unclear.

Reply:
The API must have a clear purpose. Please do not try to make it good for everything, since it may be good for nothing.

Reply:
I still believe "retrieval" from memory is the most understandable way. "get_records" is like a database operation rather than a "intelligent" memory operation.

Reply:
`retrieve` assumes a `key` argument. Here the function is parameterless. What about `summarize`?

Reply:
> I don't think letting memory tokenize context is a good idea, which is out of its responsibility. An agent memory should provide enough information and not be influenced by the front-end LLM. While the `ChatAgent` is responsible for treating overlength context. Even though we need this functionality, it should not appear in the `BaseMemory` API definition but in a specific implementation.

What is the value of this memory entity then? I need to be able to get messages and feed them into LLM. Who will be stripping messages and forming the context? Agent, okay. But why on earth I need the results of `retrieve`? Maybe they are too small for the context and I want more messages. Even worse and they are too big for the context. Why on earth I need *some* messages with which I will have pain in the ass stripping to fit the context inside an agent class.

@lightaime

Reply:
> retrieve assumes a key argument. Here the function is parameterless. What about summarize?

However, for example, `ChatAgentMemory` does not summarize anything.

> Why on earth I need some messages with which I will have pain in the ass stripping to fit the context inside an agent class.

OK, I got your idea. The memory should know how much information needed by the LLM. What I afraid is, passing tokenizer into memory will highly increase the coupling between the memory and the backend model. Instead, my suggestion is to let `ChatAgent` calculate the token number for each message and include the token number in the MemoryRecord when `write_records`. Then pass the context length as a parameter when `get_records`.

Reply:
No, it's not good. The token number of the same message is different for different LLM. It's meaningless to store it.

Reply:
We have to have some level of coupling if we what to solve the problem. And the problem here is that we do not have code to smartly compress the chat history into context so that the agent does not fail with TokenLimitExceeded error. This problem will grow larger if you introduce a variety of different memory implementations, and do not design how to resolve it.

If I am not mistaken, they purpose of this PR is to have an API to plug in a chat history augmented with a vector database. In its implementation say half of the LLM context will be filled with the recent history, the second half - with responses from the vector database. In this implementation the memory MUST know the context size and have access to tokenizer to fit the messages into the context. If you return some arbitrary messages, they may not fit into the context, and what does ChatAgent is supposed to do with this?

@lightaime please step in

Reply:
Yes, I agree with "the memory MUST know the context size". The reason I emphasize coupling is because I wish a memory can plug-and-play even if it has been used by other agents. A GPT4 agent memory learns something during some chats, then the memory can be copied and connected to a llama agent to answer questions. Yes, we can also achieve this by passing the tokenizer during retrieving, which means we only store messages without token numbers in the memory. It may be very difficult for memory to construct a proper length, we should put more effort on optimization.

Reply:
I do not mind LLM-oblivious memory, but only if it is coupled with ContextCreator class that takes memory and creates a 2048-token (as an example) length context. With ContextCreator the memory API will be different. This is what I care about is that the API is not something from the top of one's head, but something verified in action.

Reply:
Sure. So ContextCreator will act as a "bridge" between agent and memory. Agent will NEVER directly interact with Memory.

I come up with an idea. We add a field called "importance" of (0, 1] float in `MemoryRecord` during retrieving. Except the records with `importance=1` must be included in the context, any other records can be dropped to fit the context length of the model according to the `importance` value.

Reply:
Importance may be a good idea. You need to design a properly functioning API between BaseMemory and ContextCreator.